File: sorting/bubblesort.c
#include <stdio.h>

void bubblesort(int number, int* array) {
    int i, j;
    for (i = number-1; i >= 1; i--) {
        int change = 0;
        for (j = 0; j < i; j++) {
            if (array[j] > array[j+1]) {
                int temp = array[j];
                array[j] = array[j+1];
                array[j+1] = temp;
                change = 1;
            }
        }
        if (change == 0)
            return;
    }
}

// recursive bubblesort
void rec_bubblesort(int number, int* array) {
    int j;
    int change = 0;
    for (j = 0; j < number-1; j++) {
        if (array[j] > array[j+1]) { // change
            int temp = array[j];
            array[j] = array[j+1];
            array[j+1] = temp;
            change = 1;
        }
    }
    if (change != 0)
        rec_bubblesort(number-1, array);
}

int main(void) {
    int i;
    int array[8] = { 25, 48, 37, 12, 57, 86, 33, 92 };
    rec_bubblesort(8, array);
    printf("sorted array: ");
    for (i = 0; i < 8; i++)
        printf("%d ", array[i]);
    printf("\n");
    return 0;
}

File: sorting/bubblesort.ts
function swap(arr: number[], i: number, j: number) {
    [arr[i], arr[j]] = [arr[j], arr[i]];
}

function bubblesort(array: number[]) {
    for (let i = 0; i < array.length - 1; i++) {
        let swapped: boolean = false;
        for (let j = 0; j < array.length - 1; j++) {
            if (array[j] > array[j + 1]) {
                swap(array, j, j + 1);
                swapped = true;
            }
        }
        if (!swapped) break;
    }
}

const arr = [5, 3, 1, 2, 4];
bubblesort(arr);
console.log(arr);
